nix/buildLeanPackage.nix

```diff
@@ -44,11 +59,27 @@ with builtins; let
     allowSubstitutes = false;
   };
   srcRoot = src;
-  allDeps = lib.foldr (dep: allDeps: allDeps // { ${dep.name} = dep; } // dep.allDeps) {} deps;
+
+  # A flattened list of Lean-module dependencies (`deps`)
+  allExternalDeps = lib.foldr (dep: allExternalDeps: allExternalDeps // { ${dep.name} = dep; } // dep.allExternalDeps) {} deps;
+  allNonBuiltInDeps = lib.filter (dep: !(builtInDepNames ? ${dep.name})) (attrValues allExternalDeps);
+  # A flattened list of all static library dependencies: this and every dep module's explicitly provided `staticLibDeps`, 
+  # plus every dep module itself: `dep.staticLib`
+  allStaticLibDeps = 
+    lib.unique (lib.flatten (staticLibDeps ++ (map (dep: [dep.staticLib] ++ dep.staticLibDeps or []) allNonBuiltInDeps)));
+  # To build plugins, we also need to collect all static libs generated with -fPIC. Since -fPIC prevents some optimizations,
+  # we only use it for building plugins.
+  allPICStaticLibDeps = 
+    lib.unique (lib.flatten (staticLibDeps ++ (map (dep: [dep.PICstaticLib] ++ dep.PICstaticLibDeps or []) allNonBuiltInDeps)));
+  # A flattened list of all plugin dependencies
```

Comment:
Good point, made non-recursive.
____
So for some background, I'm now working on an experiment where
1) Package A defines a natively-implemented function
2) Package B uses it, and to develop it interactively I also want these definitions to be visible to the interpreter.
For 2), I also needed to minimally patch the interpreter to look up symbols in loaded plugins, but I understand that's a possibly-contentious change so I didn't want to do a PR for that yet.

So, this plugin inheritance was to help my situation. But since it's bad for linker-style plugins and these are the only ones officially supported right now, I'm changing them to be non-recursive in the PR. In the future, if upstream adds support for symbol-lookup-in-plugins, we could add a separate exportPluginRecursively attribute or something.

Reply:
> I also needed to minimally patch the interpreter to look up symbols in loaded plugins

Was this about `RTLD_GLOBAL`? We definitely want this use case to work, yes.

Reply:
For Linux yes -- just that one change -- but Windows doesn't have an equivalent of RTLD_GLOBAL so one needs to keep a list of module references and scan through them. Sg, I'll send a pr for that as well.

Reply:
What exactly do you mean by "defines a natively-implemented function"? It sounds like `[extern]`, but then loading A as a plugin wouldn't solve anything, would it...?

Reply:
> For Linux yes -- just that one change -- but Windows doesn't have an equivalent of RTLD_GLOBAL so one needs to keep a list of module references and scan through them. Sg, I'll send a pr for that as well.

Ah, you are right. That would be great, yes!

Reply:
(I hope I'm not missing something obvious!) Yup, I meant `[extern]`, more specifically `[extern]` constants. Like, in package A I have
```
@[extern c inline "*(double*)(&#1)"]
constant Float.ofUInt64Transmute (i: UInt64) : Float
```
in package B I have 

```
import A

def main : IO UInt32 := do
  IO.println $ Float.ofUInt64Transmute 123412341231231234
  return 0
```

for this to build as an executable, I need to build the objects, link them and everything works. But for development convenience, when working in package B I sometimes want to do `#eval Float.ofUInt64Transmute 5`. But `#eval` is executed by the `lean` interpreter, which looks into A's '.olean' file, finds the definition, sees that it's an `extern` constant, tries to look up the generated symbol, doesn't find it and fails. So my first idea was to make `lean` see the symbols by loading these libraries as plugins, and making lean look for symbols in plugins. Not sure if that's the design you intended :)

Reply:
I see. We've discussed loading any (transitive) dependency as a "plugin" (maybe not via that exact flag) automatically at some point, which should resolve this.